// @swagger-api/apidom-json-pointer/src/index.cjs
"use strict";

exports.__esModule = true;
exports.unescape = exports.testReferenceToken = exports.testJSONPointer = exports.testArrayLocation = exports.testArrayIndex = exports.testArrayDash = exports.parse = exports.evaluate = exports.escape = exports.compile = exports.XMLTranslator = exports.URIFragmentIdentifier = exports.JSONString = exports.JSONPointerTypeError = exports.JSONPointerParseError = exports.JSONPointerKeyError = exports.JSONPointerIndexError = exports.JSONPointerEvaluateError = exports.JSONPointerError = exports.JSONPointerCompileError = exports.Grammar = exports.CSTTranslator = exports.ASTTranslator = void 0;
var _jsonPointer = require("@swaggerexpert/json-pointer");
exports.JSONString = _jsonPointer.JSONString;
exports.URIFragmentIdentifier = _jsonPointer.URIFragmentIdentifier;
exports.parse = _jsonPointer.parse;
exports.CSTTranslator = _jsonPointer.CSTTranslator;
exports.ASTTranslator = _jsonPointer.ASTTranslator;
exports.XMLTranslator = _jsonPointer.XMLTranslator;
exports.testJSONPointer = _jsonPointer.testJSONPointer;
exports.testReferenceToken = _jsonPointer.testReferenceToken;
exports.testArrayLocation = _jsonPointer.testArrayLocation;
exports.testArrayIndex = _jsonPointer.testArrayIndex;
exports.testArrayDash = _jsonPointer.testArrayDash;
exports.compile = _jsonPointer.compile;
exports.escape = _jsonPointer.escape;
exports.unescape = _jsonPointer.unescape;
exports.Grammar = _jsonPointer.Grammar;
exports.JSONPointerError = _jsonPointer.JSONPointerError;
exports.JSONPointerParseError = _jsonPointer.JSONPointerParseError;
exports.JSONPointerCompileError = _jsonPointer.JSONPointerCompileError;
exports.JSONPointerEvaluateError = _jsonPointer.JSONPointerEvaluateError;
exports.JSONPointerTypeError = _jsonPointer.JSONPointerTypeError;
exports.JSONPointerKeyError = _jsonPointer.JSONPointerKeyError;
exports.JSONPointerIndexError = _jsonPointer.JSONPointerIndexError;
var _apidom = require("@swaggerexpert/json-pointer/evaluate/realms/apidom");
exports.evaluate = _apidom.evaluate;
